builder/googlecompute/driver_gce.go

```diff
@@ -126,11 +128,11 @@ func NewClientOptionGoogle(account *ServiceAccount, vaultOauth string, impersona
 		// 4. On Google Compute Engine and Google App Engine Managed VMs, it fetches
 		//    credentials from the metadata server.
 		//    (In this final case any provided scopes are ignored.)
-		//
-		//    Note: (4) is not usable with OSLogin on Google Compute Engine (GCE).
-		//    The GCE service account is derived separately and used instead.
```

Comment:
Could these have been a used feature ?

Reply:
I took another a look at this and it seems that I need to rework my implementation.

gceUser was introduced in https://github.com/hashicorp/packer-plugin-googlecompute/commit/95a41be763766c46a5f404e7dc63b49570f0cf40#diff-42f2cc596979842e27b1dea3a65dc4377bd631e73fab5f51acb89231ef68f829 . That snippet was put in there to acquire the email of the service account running on Google Cloud. However, this snippet shouldn't run if service account impersonation is active.

As you can see the, access token from the metadata server is missing the email field as the VMs don't have the userinfo.email scope

```
mahamed@bastion:~$ http https://oauth2.googleapis.com/tokeninfo "Authorization: Bearer $(gcloud auth print-access-token --impersonate-service-account bootstrap@REDACTED.iam
.gserviceaccount.com)"
WARNING: This command is using service account impersonation. All API calls will be executed as [bootstrap@REDACTED.iam.gserviceaccount.com].
HTTP/1.1 200 OK
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Content-Encoding: gzip
Content-Type: application/json; charset=UTF-8
Date: Tue, 14 Dec 2021 15:23:33 GMT
Expires: Mon, 01 Jan 1990 00:00:00 GMT
Pragma: no-cache
Server: ESF
Transfer-Encoding: chunked
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 0
{
    "access_type": "online",
    "aud": "111995373771645511267",
    "azp": "111995373771645511267",
    "email": "bootstrap@REDACTED.iam.gserviceaccount.com",
    "email_verified": "true",
    "exp": "1639499012",
    "expires_in": "3598",
    "scope": "https://www.googleapis.com/auth/cloud-platform https://www.googleapis.com/auth/compute https://www.googleapis.com/auth/appengine.admin https://www.googleapis.com/auth/use
rinfo.email openid",
    "sub": "111995373771645511267"
}

mahamed@bastion:~$ http https://oauth2.googleapis.com/tokeninfo "Authorization: Bearer $(http http://metadata.g
oogle.internal/computeMetadata/v1/instance/service-accounts/bastion@REDACTED.iam.gserviceaccount.com/token?
scopes=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloud-platform,https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.
email "Metadata-Flavor: Google" | jq .access_token -r)"
HTTP/1.1 200 OK
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Content-Encoding: gzip
Content-Type: application/json; charset=UTF-8
Date: Tue, 14 Dec 2021 15:19:44 GMT
Expires: Mon, 01 Jan 1990 00:00:00 GMT
Pragma: no-cache
Server: ESF
Transfer-Encoding: chunked
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 0
{
    "access_type": "online",
    "aud": "117164364499178354736",
    "azp": "117164364499178354736",
    "exp": "1639498308",
    "expires_in": "3123",
    "scope": "https://www.googleapis.com/auth/cloud-platform"
}
```

Reply:
I have reworked the logic now. gceUser function is skipped if ImpersonatedServiceAccount is set and the impersonation client option is passed in to the oauth2 service being built in this file.